### DISCUSSION TITLE
Time Keeping Within Trials

### DISCUSSION THREAD

**Discussion by mcvc968** [jspsych/jsPsych#1198] [Q&A]
Hi, 

I just wanted to ask how the plugins (particularly the audio-keyboard-response plugin) keeps track of time since the start of each trial. 

I am trying to make it such that in the first second of every trial a response is noted but does not end the trial. After that one second I want to change response-ends-trial parameter such that a response would end the trial. Any help with this would be really appreciated. 

Thanks in advance and thanks for all the help thus far.

**Comment by becky-gilbert** [comment]
Hi @mcvc968, sorry for the slow reply. Not sure if a response to this question would still be helpful, but thought I'd suggest something anyway. What I would do is make the following modifications to the [audio-keyboard-response plugin](https://github.com/jspsych/jsPsych/blob/master/plugins/jspsych-audio-keyboard-response.js):
1) the existing `response` object ([line 96](https://github.com/jspsych/jsPsych/blob/42962039af40f1ec6dfc1dadc65b4c9832dcd43e/plugins/jspsych-audio-keyboard-response.js#L96-L99)) is just set up to hold a single RT and key value, so you can modify this so that it holds _arrays_ of RTs and keys by initializing these to empty arrays instead of `null`. This way, each time a response is made, you'll be able to push the RT and key onto these arrays.:
```js
var response = {
  rt: [],
  key: []
};
```

2) When the keyboard listener is set up with [jsPsych.pluginAPI.getKeyboardResponse](https://www.jspsych.org/core_library/jspsych-pluginAPI/#jspsychpluginapigetkeyboardresponse) (starting around [line 155](https://github.com/jspsych/jsPsych/blob/42962039af40f1ec6dfc1dadc65b4c9832dcd43e/plugins/jspsych-audio-keyboard-response.js#L155-L171)), change `persist` to `true` so that the keyboard listener continues recording keyboard responses after the first response. (If `persist` is `false` then the keyboard listener will stop after the first valid response is made.)

3) In the current `after_response` function, there's a check to see whether this is the first response, and then the response is only recorded if that's the case (starting around [line 142](https://github.com/jspsych/jsPsych/blob/42962039af40f1ec6dfc1dadc65b4c9832dcd43e/plugins/jspsych-audio-keyboard-response.js#L142-L145)). You want to record all of the responses, so you can get rid of the if statement and just use this:
```javascript
response.rt.push(info.rt);
response.key.push(info.key);
```

4) Finally (and to answer your original question!), in the `after_response` function you need to know whether or not 1 second has elapsed since the audio started playing. I think what you're saying is that, if the response happens after the 1st second then the trial should end, otherwise keep waiting for more responses. Is that right? The way that the audio timing is measured is different depending on whether WebAudio or HTML5 audio is being used. However, the good news is that the keyboard listeners handle this difference in timing automatically (note that there are actually two different possible keyboard listeners - one for WebAudio and one for HTML5 audio), which means that you can use the value of `rt` that comes from the keyboard listener to check if 1 second has passed. So you can modify [this part](https://github.com/jspsych/jsPsych/blob/42962039af40f1ec6dfc1dadc65b4c9832dcd43e/plugins/jspsych-audio-keyboard-response.js#L147-L149) of the `after_response` function like this:
```javascript
if (trial.response_ends_trial && (info.rt > 1000)) {
  end_trial();
}
```
